### ISSUE TITLE
Critical images beacon JS is injected in a place where it doesn't end up running

### ISSUE THREAD

**Issue by GoogleCodeExporter** [apache/incubator-pagespeed-mod#891]
```
"
After enabling pagespeed I noticed a weired isssue with a IP Board Installation.

When using the quote button, apart from the quote I alsso get this text:

<script pagespeed_no_defer="">(function(){var 
e=encodeURIComponent,f=window,h=document,m="width",n="documentElement",p="height
",q="length",r="prototype",s="body",t="&",u="&ci=",w="&n=",x="&rd=",y=",",z="?",
A="Content-Type",B="Microsoft.XMLHTTP",C="Msxml2.XMLHTTP",D="POST",E="applicatio
n/x-www-form-urlencoded",F="img",G="input",H="load",I="oh=",J="on",K="pagespeed_
url_hash",L="url=",M=function(a,c,d){if(a.addEventListener)a.addEventListener(c,
d,!1);else if(a.attachEvent)a.attachEvent(J+c,d);else{var 
b=a[J+c];a[J+c]=function(){d.call(this);b&&b.call(this)}}};f.pagespeed=f.pagespe
ed||{};var 
N=f.pagespeed,O=function(a,c,d,b,g){this.d=a;this.f=c;this.g=d;this.a=g;this.c={
height:f.innerHeight||h[n].clientHeight||h[s].clientHeight,width:f.innerWidth||h
[n].clientWidth||h[s].clientWidth};this.e=b;this.b={}};O[r].j=function(a){a=a.ge
tBoundingClientRect();return{top:a.top+(void 
0!==f.pageYOffset?f.pageYOffset:(h[n]||h[s].parentNode||h[s]).scrollTop),left:a.
left+(void 
0!==f.pageXOffset?f.pageXOffset:(h[n]||h[s].parentNode||h[s]).scrollLeft)}};O[r]
.i=function(a){if(0>=a.offsetWidth&&0>=a.offsetHeight)return!1;a=this.j(a);var 
c=a.top.toString()+y+a.left.toString();if(this.b.hasOwnProperty(c))return!1;this
.b[c]=!0;return a.top<=this.c[p]&&a.left<=this.c[m]};O[r].l=function(){for(var 
a=[F,G],c=[],d={},b=0;b{d=y+e(c[b]);if(131072=a[q]+d[q]&&(a+=d),b=!0);N.critical
ImagesBeaconData=a;if(B){var c=this.d,b=this.f,l;if(f.XMLHttpRequest)l=new 
XMLHttpRequest;else if(f.ActiveXObject)try{l=new 
ActiveXObject(C)}catch(P){try{l=new 
ActiveXObject(B)}catch(Q){}}l&&(l.open(D,c+(-1==c.indexOf(z)?z:t)+L+e(B)),l.setR
equestHeader(A,E),l.send(a))}};O[r].h=function(){for(var 
a={},c=h.getElementsByTagName(F),d=0;d=a[b.src].n&&b[p]>=a[b.src].m)a[g]={render
edWidth:b[m],renderedHeight:b[p],originalWidth:b.naturalWidth,originalHeight:b.n
aturalHeight}}return a};N.k=function(a,c,d,b,g){var k=new 
O(a,c,d,b,g);M(f,H,function(){f.setTimeout(function(){k.l()},0)})};N.criticalIma
gesBeaconInit=N.k;})();pagespeed.criticalImagesBeaconInit('/ngx_pagespeed_beacon
','https://forums.plex.tv/index.php?s=7c52f3b4f6a969052159ac848c76a011&app=forum
s&module=ajax§ion=topics&do=quote&t=59548&p=353044&md5check=ca71f5bc3a13296b514
46351e109b37e&isRte=1','W2PyaY9kww',false,'8ByaY1t0dHM');</script>
"

Out of curiosity, what tags (if any) occur after the script on your page?  It 
looks like we're injecting the script in the wrong place and that's what's 
causing the trouble, and I'd like to diagnose why this might be.


That's what I see in the inspector... do you want me to expand the div or is 
that enough?

    <!DOCTYPE html>
    <html lang="en">
        <head></head>
        <body id="ipboard_body" onpageshow="event.persisted && (function(){var allInstances = CKEDITOR.i… doc.$.designMode = "off"; doc.$.designMode = "on"; }}})();">
            <p id="content_jump" class="hide"></p>
            <div id="ipbwrapper"></div>
            <script type="text/javascript"></script>
            <script pagespeed_no_defer=""></script>   <--- This is the script in question
            <div id="overlay" style="display: none;"></div>
            <div id="lightbox" style="display: none;"></div>
            <input id="hte_editor_52fd475215632" type="text" style="position: absolute; left: -1000px; display: none;" undefined="position:absolute;left:-1000px;display:none;"></input>
            <div id="ajax_loading" style="display: none;"></div>
        </body>
    </html>
```

Original issue reported on code.google.com by `jmaes...@google.com` on 18 Feb 2014 at 4:33

**Comment by sachinjsk** [comment]
I ran into the same issue. Setting 'ModPagespeedCriticalImagesBeaconEnabled false' did the trick for me.

**Comment by jlporter** [comment]
Do you have a page or URL that demonstrates the problem? That would help a lot in tracking down the issue.

**Comment by sachinjsk** [comment]
@jlporter http://www.babyimage4d.sbi1b.sitesell.com/ . You'll notice the JS code appearing for few seconds before the slider at the top right corner is rendered.

Do you think turning ModPagespeedCriticalImagesBeaconEnabled off will have any adverse effect? Documentation says "If you disable image beacons but enable filters that use them, the filters will work but not as well as when beacons are enabled.". But its not so clear what the implications are. Would you know?

**Comment by jlporter** [comment]
Thanks for the test case, that's very helpful and I see what the problem is. The critical image beacon JS gets injected immediately before the first image on the page. On your page, the first image is in the slideshow div. There is a bit of JS that runs on your site to cycle through the various images by setting `display: block` on each image in the slideshow in succession. This slideshow JS is also applying `display: block` to the JS snippet, which is causing it to be displayed. Is it possible for you to update your JS to ignore the script tag being injected into the slideshow?

To your other question, the main benefit of the critical images beacon is in improving the behavior of the [lazyload images filter](https://developers.google.com/speed/pagespeed/module/filter-lazyload-images). When beaconing is enabled, the above the fold images will not be lazy loaded, improving the perceived loading time of the page. It can be a nice win, but like so many of these things, depends a lot on the specific properties and layout of your page. You can read a bit more about it [here](https://developers.google.com/speed/pagespeed/module/filter-lazyload-images#beacons).

**Comment by jlporter** [comment]
One other benefit of the image beaconing I neglected to mention is that it can improve the behavior of [inline_preview_images](https://developers.google.com/speed/pagespeed/module/filter-inline-preview-images) and [inline_images](https://developers.google.com/speed/pagespeed/module/filter-image-optimize). When beaconing is used with these images, only the above-the-fold images will be inlined. This allows the images at the top of the page to be displayed quickly, while keeping the images at the bottom of the page outlined to prevent bloating up the size of the HTML.

**Comment by sachinjsk** [comment]
@jlporter Thanks for looking into this issue. I see what's going on. That's a client site and I don't have direct access to edit the source. 

I'm rolling out Pagespeed to few thousand client sites, so I'll need to figure out a more generic solution. I'm sure at least few of them will have some sort of similar slider plugin. Right now I'm rolling out with ModPagespeedCriticalImagesBeaconEnabled turned off. Would be great if we could find a generic solution.

**Comment by jeffkaufman** [comment]
Fixed by c6f01f2
